biogtr/io/association_matrix.py

```diff
@@ -102,61 +114,97 @@ def to_dataframe(
+            if len(col_labels) == len(self.ref_instances):
+                col_inds = col_labels
+            else:
+                raise ValueError(
+                    (
+                        f"Mismatched # of columns and labels!",
+                        f"Found {len(col_labels)} with {len(self.ref_instances)} columns",
+                    )
+                )
         else:
-            col_inds = np.arange(len(self.ref_instances))
+            if col_labels.lower() == "gt":
+                col_inds = [
+                    instance.gt_track_id.item() for instance in self.ref_instances
+                ]
+
+            elif col_labels.lower() == "pred":
+                col_inds = [
+                    instance.pred_track_id.item() for instance in self.ref_instances
+                ]
+
+            else:
+                col_inds = np.arange(len(self.ref_instances))
 
         asso_df = pd.DataFrame(matrix, index=row_inds, columns=col_inds)
 
         return asso_df
 
     def reduce(
         self,
-        to: tuple[str] = ("inst", "traj"),
-        by: tuple[str] = (None, "pred"),
+        row_dims: str = "inst",
+        col_dims: str = "traj",
+        row_grouping: str = None,
+        col_grouping: str = "pred",
         reduce_method: callable = np.sum,
     ) -> pd.DataFrame:
-        """Reduce association matrix rows/columns to inst/traj x traj/inst.
+        """Aggregate the association matrix by specified dimensions and grouping.
 
         Args:
-           to: A tuple indicating how to reduce rows/columns. Either inst (remains unchanged), or traj (reduces matrix from n_query/n_ref to n_traj
-           by: A tuple indicating how to group rows/columns when aggregating. Either "pred" or "gt"
+           row_dims: A str indicating how to what dimensions to reduce rows to. Either inst (remains unchanged), or traj (n_rows=n_traj)
+           col_dims: A str indicating how to dimensions to reduce rows to. Either inst (remains unchanged), or traj (n_cols=n_traj)
+           row_grouping: A str indicating how to group rows when aggregating. Either "pred" or "gt".
+           col_grouping: A str indicating how to group columns when aggregating. Either "pred" or "gt".
            method: A callable function that operates on numpy matrices and can take an `axis` arg for reducing.
 
         Returns:
             The association matrix reduced to an inst/traj x traj/inst association matrix as a dataframe.
         """
-        row_to, col_to = to
-        row_by, col_by = by
-
-        if row_to not in ("inst", "traj") or col_to not in ("inst", "traj"):
+        if (
+            row_dims is not None
+            and row_dims.lower()
+            not in [key for keys in self.AVAILABLE_REDUCTIONS.values() for key in keys]
+        ) or (
+            col_dims is not None
+            and col_dims.lower()
+            not in [key for keys in self.AVAILABLE_REDUCTIONS.values() for key in keys]
```

Comment:
Cache all of this. Either have pre-defined lists (run this double loop comprehension at the module level), or have a helper function with a `@functools.cache` decorator. The former is easier and does the job. I'd even recommend storing it in a tuple or set, depending on whether order matters, to make it faster to query.